packages/replayer/src/debugger/replay-debugger.ui.ts

```diff
@@ -0,0 +1,175 @@
+import { ReplayableEvent } from "@alwaysmeticulous/api";
+import { startServer } from "@alwaysmeticulous/replay-debugger-ui";
+import { BeforeUserEventOptions } from "@alwaysmeticulous/sdk-bundles-api";
+import { Browser, Page } from "puppeteer";
+
+export interface ReplayDebuggerState {
+  events: ReplayableEvent[];
+  index: number;
+  loading: boolean;
+}
+
+export interface ReplayDebuggerUIOptions {
+  browser: Browser;
+  replayedPage: Page;
+  replayableEvents: ReplayableEvent[];
+}
+
+export interface ReplayDebuggerUI {
+  page: Page;
+}
+
+type OnBeforeUserEventCallback = (
+  options: BeforeUserEventOptions
+) => Promise<void>;
+
+export const openStepThroughDebuggerUI = async ({
+  browser,
+  replayedPage,
+  replayableEvents,
+}: ReplayDebuggerUIOptions): Promise<OnBeforeUserEventCallback> => {
+  // Start the UI server
+  const uiServer = await startServer();
+
+  // Create page for the debugger UI
+  const debuggerPage = await browser.defaultBrowserContext().newPage();
+  debuggerPage.setViewport({
+    width: 0,
+    height: 0,
+  });
+
+  /**
+   * The index the page is in the process of advancing to. Equal to the current index
+   * if the page has already replayed all the so-far-requested user events.
+   */
+  let targetIndex = 0;
+  let state: ReplayDebuggerState = {
+    events: replayableEvents,
+    index: 0,
+    loading: false,
+  };
+
+  const setState = async (newState: Partial<ReplayDebuggerState>) => {
+    state = { ...state, ...newState };
+    await debuggerPage.evaluate((state) => {
+      (window as any).__meticulous__replayDebuggerSetState(state);
+    }, state as any);
+  };
+
+  const onReady = async () => {
+    await setState(state);
+  };
+
+  let advanceToNextEvent: (() => void) | null = null;
+  const onBeforeNextEvent = async ({
+    userEventIndex,
+  }: BeforeUserEventOptions) => {
+    await setState({
+      loading: userEventIndex < targetIndex,
+      index: Math.max(0, Math.min(state.events.length - 1, userEventIndex)),
+    });
+
+    if (state.index < targetIndex) {
+      advanceToNextEvent = null;
+      return; // keep going
+    }
+
+    return new Promise<void>((resolve) => {
+      advanceToNextEvent = resolve;
+    });
+  };
+
+  const onPlayNextEvent = async () => {
+    targetIndex = state.index + 1;
+    await setState({ loading: true });
+    advanceToNextEvent?.();
+  };
+
+  const onAdvanceToIndex = async (newTargetIndex: number) => {
+    if (newTargetIndex <= state.index) {
+      return; // Do nothing
+    }
+    targetIndex = newTargetIndex;
+    await setState({ loading: true });
+    advanceToNextEvent?.();
+  };
+
+  const findEventTarget = async (event: ReplayableEvent) => {
```

Comment:
This and onCheckNextTarget are unchanged, just moved over from the old replay-debugger.ui.ts file